### ISSUE TITLE
Starting with version 2.0.182 an error occurs when calva loads in WSL

### ISSUE THREAD

**Issue by vamem9z-Moses** [BetterThanTomorrow/calva#1097]
When opening a project that is located in the WSL environment starting with version 2.0.182 the following errors show up in the output window.

[Error - 6:13:32 PM] Connection to server is erroring. Shutting down server.
[Error - 6:13:32 PM] Connection to server is erroring. Shutting down server.
[Error - 6:13:32 PM] Connection to server is erroring. Shutting down server.
[Error - 6:13:32 PM] Connection to server is erroring. Shutting down server.
/home/mmiles/.vscode-server/extensions/betterthantomorrow.calva-2.0.184/clojure-lsp: 1: Syntax error: "(" unexpected

Is there something else I can try?

**Comment by PEZ** [comment]
Thanks for reporting!

This was mention on Slack as well, even though I think we didn't make the connecting to version .182 then. Maybe we need a fallback on a bundled jar, @bpringe ?

**Comment by bpringe** [comment]
Maybe so! I'll prioritize this.

**Comment by bpringe** [comment]
@vamem9z-Moses Can you post the contents of the developer console after starting VS Code? `Help -> Toggle Developer Tools -> Console tab`. There should be some info about the starting of clojure-lsp, and what things did or didn't happen during that time.

**Comment by vamem9z-Moses** [comment]
@bpringe I've attached the startup log. I noticed this Extension Host commnet - Downloading clojure-lsp from /clojure-lsp/clojure-lsp/releases/download/2021.03.30-20.42.34/clojure-lsp-native-linux-amd64.zip. 

I'm using WSL 2 on a Surface Pro X so the OS is Ubuntu 20.04 on aarch64. I'm not sure if that makes a difference.  I don't encounter this error if I downgrade to Calva version 2.0.181.
[vscode calva 2.0.184 start up log.txt]
(https://github.com/BetterThanTomorrow/calva/files/6251904/vscode.calva.2.0.184.start.up.log.txt)

**Comment by bpringe** [comment]
Yeah, it looks like the binary is not compatible with your OS, so the jar should be used. I'll need to think on this a bit. We could bundle a jar with Calva, but then not only would we have to keep downloading new versions manually to release with Calva, but the setting for the user to change the lsp version would not work for systems using the jar.

I'll see if we can detect the error, then download and use the jar, and also make this method work the same as the binary method in terms up only downloading a given version once, etc.

**Comment by PEZ** [comment]
How about a setting to use binary or jar?

**Comment by bpringe** [comment]
That's a good idea. I'd like to first see if we can easily detect the error, and if so, maybe we can avoid adding a setting, since so far the only reason I could imagine someone would want to change that setting to use the jar is if the binary isn't working. However, the difficulty with _that_ is that the flow becomes:

1. Download binary and run it.
2. On error, download jar and run it.
(easy enough up until now)
3. Now... we need to make sure the download of the binary doesn't happen again in the future. So we need to save somewhere that on this machine we should always check for and download the jar, rather than the binary.

So, having typed out the above, it seems that number 3 is basically the same as having a setting, except we set it ourselves for all future starts, and there's no easy way to _unset_ it, leading right back to your idea of adding a user setting. :smile:

@ericdallo, what does lsp-mode do/have to mitigate this issue (the binary does not work for the OS)? I think you mentioned it allows setting a path to clojure-lsp.

I wonder if Calva should add that. Currently there's the static-linux binary that I'm guessing is the only one that would work for something like Nix (disregarding the jar)? In this case, with Calva, there's no way for someone to use that with Calva.

**Comment by ericdallo** [comment]
Really odd error, could you share more info of your OS, it's really odd that exception

> @ericdallo, what does lsp-mode do/have to mitigate this issue (the binary does not work for the OS)? I think you mentioned it allows setting a path to clojure-lsp.

Yeah, lsp-mode accept an optional path to the executable [here](https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-clojure.el#L39)

@bpringe maybe you can allow use the deprecated `clojure-lsp` embedded jar from the release, but my idea was to remove it in the future as we didn't had issues with graalvm until now 😅

**Comment by vamem9z-Moses** [comment]
@ericdallo  Thanks for looking into this.  Is there any chance of getting an linux arm64 native image of clojure-lsp.? As of version 21.0.0 graalvm has full experimental support for the arm64 architecture. 

**OS INFO**

Here is the information for the host os.  

Edition	Windows 10 Home
Version	20H2
Installed on	‎1/‎10/‎2021
OS build	19042.870
Serial number	035543194953
Experience	Windows Feature Experience Pack 120.2212.551.0

Processor	Microsoft SQ1 @ 3.0 GHz   3.00 GHz
Installed RAM	15.9 GB (15.5 GB usable)
Product ID	00325-80000-00000-AAOEM
System type	64-bit operating system, ARM-based processor
Pen and touch	Pen and touch support with 10 touch points

Here is the info for the WSL environment

5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:36:35 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux (Ubuntu 20.04)

**Comment by ericdallo** [comment]
Feel free to open an issue on clojure-lsp side to support that, We need to check how to generate the arm64 binary via GH actions just like the other binaries, also test if everything work like sqlite, but the integration tests should help most of the testing

**Comment by ericdallo** [comment]
I added a experimental CI process for that [here](https://github.com/clojure-lsp/clojure-lsp/commit/e16b471d26943ee3edca4bf4a3f6bad3b25ee760) but I'm not sure it'll work yet as GH actions don't have a aarch64 runner yet

**Comment by bpringe** [comment]
> maybe you can allow use the deprecated clojure-lsp embedded jar from the release, but my idea was to remove it in the future as we didn't had issues with graalvm until now

Oh, well I'll wait and see if clojure-lsp can support that OS. In any case, I don't want to make Calva rely on a deprecated artifact.

**Comment by ericdallo** [comment]
Yeah, I'll need to make some tests as well, GH actions seem to not support running in aarch64 yet, so we will need to do some manual tests if the generated binary work well on aarch64

**Comment by vamem9z-Moses** [comment]
Let me know if I can help test anything.

**Comment by somanythings** [comment]
I think I might be suffering the same or similar bug. I noticed just a few days ago though that my vscode in wsl2 was reporting "Connection to server is erroring. Shutting down server" with 2.0.211, but I'm not getting an error cause like that syntax error in the OP. I've just tried downgrading to 2.0.181, and the after a couple of window restarts the error is gone. I'd tried delete the .lsp dirs in the project directory, but no difference. 
Around the same time I started seeing messages about missing java language runtime. I've had no luck setting that up for wsl2. -

_That error also seems to be gone now I've downgraded to 2.0.181_ [Edit: no it isn't, but I didn't see it until a restart, rather than just a window reload, i think]

Let me know if I can add any further logs to help diagnose. Also happy to test.

**Comment by bpringe** [comment]
> We need to check how to generate the arm64 binary via GH actions just like the other binaries, also test if everything work like sqlite

@ericdallo Any updates on this, or ideas for other ways to make this work for them?

**Comment by bpringe** [comment]
A possible workaround (I'm unsure if this would work), is to download the binary just called `clojure-lsp` from the clojure-lsp releases, which contains and runs a jar, from what I understand. https://github.com/clojure-lsp/clojure-lsp/releases/tag/2021.09.13-22.25.35

You can then set Calva to run this downloaded binary: https://calva.io/clojure-lsp/#using-a-custom-clojure-lsp-native-binary

**Comment by ericdallo** [comment]
> @ericdallo Any updates on this, or ideas for other ways to make this work for them?

Not really, I'm planning drop sqlite soon and then we can probably make that work: https://github.com/clojure-lsp/clojure-lsp/issues/549

> A possible workaround (I'm unsure if this would work), is to download the binary just called clojure-lsp from the clojure-lsp releases, which contains and runs a jar, from what I understand. https://github.com/clojure-lsp/clojure-lsp/releases/tag/2021.09.13-22.25.35

Correct

**Comment by somanythings** [comment]
Thanks, I went to go try this, I'd downloaded it, as was going to make the clojure-lsp path setting, when, much to my surprise, things are working again!? I'm back on calva v2.0.211, 

And now get this output: 
```
Calva is utilizing cider-nrepl and clojure-lsp to create this VS Code experience.
  nREPL dependencies configured:
    nrepl: 0.8.3
    cider-nrepl: 0.26.0
    cider/piggieback: 0.5.2
 clojure-lsp version configured: latest

If you are new to Calva, please consider starting with the command:
  **Calva: Fire up the Getting Started REPL**
  https://calva.io/getting-started/

(See `showCalvaSaysOnStart` in Settings to control the auto-showing of this message panel.)

clojure-lsp version used: 2021.09.13-22.25.35
clj-kondo version used: 2021.08.07-SNAPSHOT

```

I'm not sure what I might have changed. Possibly a machine full restart, but I thought I'd already gone through that. 🤷‍♂️
I DO still get 
![image](https://user-images.githubusercontent.com/6841417/134764564-4f3eafb1-2984-4011-9090-3b3c1ab1b591.png)
but I'm not sure if that's causing any problems with calva and clojure. It doesn't seem to like any path I give it to a jdk, whether on windows side, or wsl2/ubuntu side.

**Comment by PEZ** [comment]
Happy it is working! Maybe some file locking or similar needed help to clear up...

The Java 11 error is strictly something about the Java extension. Google it in isolation from Clojure and Calva and maybe you can find some clues.

**Comment by bpringe** [comment]
@vamem9z-Moses Are you still having this issue with the latest Calva and clojure-lsp?

@ericdallo Has any progress been made toward an arm64 binary?

**Comment by ericdallo** [comment]
@bpringe A little bit, since clojure-lsp is not using sqlite anymore replaced by datalevin, we have now a hope to have that working as soon datalevin has support for it which is not hard.
From datalevin's maintainer, we just need to add that support via github action: https://clojurians.slack.com/archives/C01RD3AF336/p1636381154017600
